src/main/java/org/opentripplanner/middleware/controllers/api/UserController.java

```diff
@@ -0,0 +1,241 @@
+ */
+public class UserController extends ApiController<User> {
+    private static final Logger LOG = LoggerFactory.getLogger(UserController.class);
+    private static ObjectMapper mapper = new ObjectMapper();
+    private static final String AUTH0_DOMAIN = getConfigPropertyAsText("AUTH0_DOMAIN");
+    public static final int TEST_AUTH0_PORT = 8089;
+    public static final String TEST_AUTH0_DOMAIN = String.format("localhost:%d", TEST_AUTH0_PORT);
+    public static final String DEFAULT_BASE_USERS_URL = "https://" + AUTH0_DOMAIN  + USERS_API_PATH;
+    /** Users URL uses Auth0 domain by default, but can be overridden with {@link #setBaseUsersUrl(String)} for testing. */
+    private static String baseUsersUrl = DEFAULT_BASE_USERS_URL;
+
+    public UserController(String apiPrefix){
+        super(apiPrefix, Persistence.users);
+    }
+
+    /**
+     * Before creating/storing a user in MongoDB, create the user in Auth0 and update the {@link User#auth0UserId}
+     * with the value from Auth0.
+     */
+    @Override
+    User preCreateHook(User user, Request req) {
+        validateUser(user, req);
+        return user;
+    }
+
+    private User validateUser(User user, Request req) {
+        // Ensure no user with email exists in MongoDB.
+        User userWithEmail = Persistence.users.getOneFiltered(eq("email", user.email));
+        if (userWithEmail != null) {
+            logMessageAndHalt(req, 400, "User with email already exists in database!");
+        }
+        boolean isDataToolsUser = false;
+        // Check for user in Auth0.
+        Auth0UserProfile auth0UserProfile = getUserByEmail(user.email);
+        if (auth0UserProfile != null) {
+            // If a user with email exists in Auth0, assign Auth0 ID to new user record in MongoDB.
+            // TODO: Check app_metadata on user profile?
+            isDataToolsUser = true;
+            LOG.warn("User {} already exists in Auth0. Storing new record in Mongo.", auth0UserProfile.email);
+        } else {
+            LOG.info("No user found in Auth0. Creating new one.");
+            // Otherwise, create the Auth0 user.
+            auth0UserProfile = createAuth0User(user, req);
+        }
+        if (!assignAdminRoleToUser(auth0UserProfile.user_id)) {
+            logMessageAndHalt(req, 500, "Failed to assign admin role to user");
+        }
+        LOG.info("Created user: {}", auth0UserProfile.user_id);
+        user.auth0UserId = auth0UserProfile.user_id;
+        user.isDataToolsUser = isDataToolsUser;
+        return user;
+    }
+
+    @Override
+    User preUpdateHook(User user, Request req) {
+        // TODO: Prevent email address change, prevent auth0 id change,
+        return user;
+    }
+
+    /**
+     * Before deleting the user in MongoDB, attempt to delete the user in Auth0.
+     */
+    @Override
+    boolean preDeleteHook(User user, Request req) {
+        HttpRequest.Builder deleteUserRequestBuilder = HttpRequest.newBuilder();
```

Comment:
There should be a check for whether the user making the request is authorized to delete the other user.